Threading fix when populating the default library context.

paulidale opened the pull request:
I'm not convinced that property definition locking is the right solution.
The test change fixes what was almost certainly a typo.

- [ ] documentation is added or updated
- [x] tests are added or updated

Fixes #14720

XuanThuLe commented:
Ok

t8m commented on crypto/property/defn_cache.c:
```diff
@@ -98,25 +120,31 @@ int ossl_prop_defn_set(OSSL_LIB_CTX *ctx, const char *prop,
     OPENSSL_free(p);
-    return 0;
+    res = 0;
+ fin:
```
We usually use end, but OK. :grin:

paulidale replied:
Changed.

t8m commented:
This LGTM, but probably @levitte or @mattcaswell should look at it as well.

paulidale commented:
I'd still like to have a chat with @levitte about better fixes.  This entails not making the changes to `defn_cache.c` and doing something elsewhere.

paulidale commented:
This needs reapproval because the dodgy add locking at the property definition level has been removed and in it's place I've added selective locking of the lib ctx.  The first two commits remain unchanged and are general improvements to the code.  The old third commit has been removed and replaced with a new leaner and better fix, one that locks the library context at just the right times.

paulidale commented:
Merged to master, thanks for the feedback.
